Nama : Sakinah Yunus

NIM  : H071191046

Tugas 1 Machine Learning

Import Library
Yang pertama adalah pandas sebagai pd
Kemudian mlxtend melakukan import fungsi apriori dan association_rules

In [15]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

df = pd.read_excel('Online Retail.xlsx')
df.head()

FileNotFoundError: ignored

Cleaning data dengan menghilangkan beberapa spasi, menghapus variabel invoiceNo yang kosong, dan menghilangkan transaksi yang merupakan transaksi kredit (diawali dengan huruf C pada invoice number)

In [5]:
df['Description'] = df['Description'].str.strip()
df.dropna(axis=0, subset=['InvoiceNo'], inplace=True)

df['InvoiceNo'] = df['InvoiceNo'].astype('str') 
df = df[~df['InvoiceNo'].str.contains('C')]

Membuat keranjang belanja yang dikenali berdasarkan nomor invoice yang berasal dari negara Perancis saja. 

In [9]:
basket = (df[df['Country'] == "France"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))
basket.head()

Description  10 COLOUR SPACEBOY PEN  ...  ZINC T-LIGHT HOLDER STARS SMALL
InvoiceNo                            ...                                 
536370                          0.0  ...                              0.0
536852                          0.0  ...                              0.0
536974                          0.0  ...                              0.0
537065                          0.0  ...                              0.0
537463                          0.0  ...                              0.0

[5 rows x 1563 columns]

Penjelasan output diatas :
Jika jumlah barang ada, maka nilai 0 akan diganti sesuai jumlah barang pada masing-masing invoice (nomor nota).

In [10]:
def encode_units(x) :
  if x <= 0:
    return 0
  if x >= 1:
    return 1
basket_sets = basket.applymap(encode_units)

#menghapus kolom postage
basket_sets.drop('POSTAGE', inplace=True, axis=1)
basket_sets.head()

Description  10 COLOUR SPACEBOY PEN  ...  ZINC T-LIGHT HOLDER STARS SMALL
InvoiceNo                            ...                                 
536370                            0  ...                                0
536852                            0  ...                                0
536974                            0  ...                                0
537065                            0  ...                                0
537463                            0  ...                                0

[5 rows x 1562 columns]

Kemudian melakukan encoding, dimana jika barang kurang dari sama dengan 0 maka keranjang tersebut bernilai 0 dan jika lebih dari 1 maka nilainya adalah 1, sehingga jika sebuah nota membeli barang A sebanyak 10 buah maka hanya akan dihitung 1. 

Selanjutnya adalah membuat variabel dimana terdiri dari beberapa barang yang sering / terbeli dari seluruh transaksi menggunakan perintah apriori

In [12]:
# Build up the frequent items 
frequent_itemsets = apriori(basket_sets, min_support=0.07, use_colnames=True)
frequent_itemsets.head()

support                        itemsets
0  0.071429   (4 TRADITIONAL SPINNING TOPS)
1  0.096939    (ALARM CLOCK BAKELIKE GREEN)
2  0.102041     (ALARM CLOCK BAKELIKE PINK)
3  0.094388      (ALARM CLOCK BAKELIKE RED)
4  0.081633  (BAKING SET 9 PIECE RETROSPOT)

Dimana perintah yang di gunakan adalah apriori, dengan data dari basket_sets dengan minimum nilai support 0.07/ 7%.

In [13]:
# Create the rules
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules

antecedents  ... conviction
0                         (ALARM CLOCK BAKELIKE PINK)  ...   3.283859
1                        (ALARM CLOCK BAKELIKE GREEN)  ...   3.791383
2                          (ALARM CLOCK BAKELIKE RED)  ...   5.568878
3                        (ALARM CLOCK BAKELIKE GREEN)  ...   4.916181
4                          (ALARM CLOCK BAKELIKE RED)  ...   4.153061
5                         (ALARM CLOCK BAKELIKE PINK)  ...   3.293135
6                                (SPACEBOY LUNCH BOX)  ...   2.101190
7                              (DOLLY GIRL LUNCH BOX)  ...   3.102273
8                     (PLASTERS IN TIN CIRCUS PARADE)  ...   1.835747
9                          (PLASTERS IN TIN SPACEBOY)  ...   2.363588
10                    (PLASTERS IN TIN CIRCUS PARADE)  ...   2.104592
11                 (PLASTERS IN TIN WOODLAND ANIMALS)  ...   2.063681
12                 (PLASTERS IN TIN WOODLAND ANIMALS)  ...   2.221939
13                         (PLASTERS IN TIN SPACEBOY)  ...   3.443878
14               (SET/20 RED RETROSPOT PAPER NAPKINS)  ...   3.736395
15                      (SET/6 RED SPOTTY PAPER CUPS)  ...   3.345481
16               (SET/20 RED RETROSPOT PAPER NAPKINS)  ...   3.780612
17                    (SET/6 RED SPOTTY PAPER PLATES)  ...   4.336735
18                      (SET/6 RED SPOTTY PAPER CUPS)  ...   7.852041
19                    (SET/6 RED SPOTTY PAPER PLATES)  ...  21.556122
20  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.897959
21  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.489796
22  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...  ...   4.625850
23               (SET/20 RED RETROSPOT PAPER NAPKINS)  ...   3.510204
24                      (SET/6 RED SPOTTY PAPER CUPS)  ...   3.232653
25                    (SET/6 RED SPOTTY PAPER PLATES)  ...   4.081633

[26 rows x 9 columns]

Selanjutnya adalah membangun sebuah variabel yang memiliki aturan aturan asosiasi dari masin-masing barang, variabel rulesmerupakan hasil dari fungsi yang mencari asosiasi dimana data yang di gunakan berasal dari frequent_items , dengan nilai minium dari lift ratio nya adalah 1

Bisa dilihat pada output, barang Alarm clock bakelike green biasanya terbeli bersamaan dengan Alarm Clock bakelike pink dengan nilai support 0.096939.

Kemudian selanjutnya adalah melakukan filter untuk nilai lift ratio minimal 6 dan tingkat confidence minimal 0.8

In [14]:
rules[ (rules['lift'] >= 6) &
       (rules['confidence'] >= 0.8) ]

antecedents  ... conviction
2                          (ALARM CLOCK BAKELIKE RED)  ...   5.568878
3                        (ALARM CLOCK BAKELIKE GREEN)  ...   4.916181
17                    (SET/6 RED SPOTTY PAPER PLATES)  ...   4.336735
18                      (SET/6 RED SPOTTY PAPER CUPS)  ...   7.852041
19                    (SET/6 RED SPOTTY PAPER PLATES)  ...  21.556122
20  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.897959
21  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.489796
22  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...  ...   4.625850

[8 rows x 9 columns]